### ISSUE TITLE
createdb often starts before the database is truly ready

### ISSUE THREAD

**Issue by ocharles** [jfischoff/tmp-postgres#9]
See this output:

```
The files belonging to this database system will be owned by user "ollie".
This user must also own the server process.

The database cluster will be initialized with locales
  COLLATE:  en_US.UTF-8
  CTYPE:    en_GB.UTF-8
  MESSAGES: en_US.UTF-8
  MONETARY: en_US.UTF-8
  NUMERIC:  en_US.UTF-8
  TIME:     en_US.UTF-8
The default text search configuration will be set to "english".

Data page checksums are disabled.

creating directory /tmp/tmp-postgres11933/data ... ok
creating subdirectories ... ok
selecting default max_connections ... 100
selecting default shared_buffers ... 128MB
selecting dynamic shared memory implementation ... posix
creating configuration files ... ok
running bootstrap script ... ok
performing post-bootstrap initialization ... ok
syncing data to disk ... ok

Success. You can now start the database server using:

    /home/ollie/.nix-profile/bin/pg_ctl -D /tmp/tmp-postgres11933/data -l logfile start

LOG:  connection received: host=[local]
LOG:  incomplete startup packet
LOG:  database system was shut down at 2017-10-19 10:17:28 GMT
LOG:  connection received: host=[local]
FATAL:  the database system is starting up
LOG:  MultiXact member wraparound protections are now enabled
LOG:  connection received: host=[local]
FATAL:  the database system is starting up
createdb: could not connect to database template1: FATAL:  the database system is starting up
LOG:  database system is ready to accept connections
LOG:  autovacuum launcher started
LOG:  received fast shutdown request
LOG:  aborting any active transactions
LOG:  autovacuum launcher shutting down
LOG:  shutting down
LOG:  database system is shut down
query-tests: CreateDBFailed (ExitFailure 1)
```

Repeatedly running this command does eventually work. Notice that `createdb` has failed before the output `database system is ready`.

**Comment by ocharles** [comment]
https://www.postgresql.org/docs/9.3/static/app-pg-isready.html has been around since 9.3, could we require that?

**Comment by jfischoff** [comment]
@ocharles what does `app-pg-isready` besides attempting to connect to the db? The code at least attempts to connect the db before calling createdb: https://github.com/jfischoff/tmp-postgres/blob/master/src/Database/Postgres/Temp/Internal.hs#L178

My understanding is postgres refuses connections before it is ready, but perhaps that is untrue or we are not properly checking for a successful connection?

**Comment by ocharles** [comment]
`waitForDB` doesn't actually wait for a valid connection, just that the socket could be opened. If you see the log above, the exception is thrown when you try and run `createdb`:

```
createdb: could not connect to database template1: FATAL:  the database system is starting up
```

`pg_isready` actually waits for the database to be ready.

**Comment by jfischoff** [comment]
> waitForDB doesn't actually wait for a valid connection, just that the socket could be opened. 
I'm not sure I follow. From my reading `waitForDB` opens a socket and tries to `connect`. As I understand it the `connect` should block.

https://github.com/jfischoff/tmp-postgres/blob/master/src/Database/Postgres/Temp/Internal.hs#L30-L33

If the `connect` fails it retries. 

> If you see the log above, the exception is thrown when you try and run createdb:

I understand that. I read your report and I am not denying that this happening. I'm just not sure I understand why `waitDB` is not sufficient. 

Looking through the source code of `pg_isready` it calls postgres client connection functions. If you  follow the connection code it ultimately makes the connection here: https://github.com/postgres/postgres/blob/db6986f47c9531628d151d6bf760a2fe1214b19d/src/interfaces/libpq/fe-connect.c#L2242

**Comment by ocharles** [comment]
> I understand that. I read your report and I am not denying that this happening. I'm just not sure I understand why waitDB is not sufficient.

Because just opening a socket and calling `connect` is not sufficient. That connection is not ready to accept PostgreSQL commands if the database is still starting up - hence the fatal exception.

> Anyway, I don't think we need to use pg_isready. Any postgres client should be sufficient. I think the change here is just to use the postgresql-simple connections function instead of networks socket connection call.

Maybe, I'll leave that to your discretion. That said, I suggested `pg_isready` because it's doing exactly what you want - to wait until a connection is ready 😄

**Comment by jfischoff** [comment]
I see what you are saying now, and we are in agreement :).

My rationale for not using pg_isready is some postgres os packages drop
standard exes so the fewer i have to depend on the better.

**Comment by jfischoff** [comment]
Should be fixed in bfbc639
